providers/implementations/storemgmt/file_store.c

```diff
@@ -773,6 +774,46 @@ static int file_close(void *loaderctx)
     return 1;
 }
 
+static void *file_open_ex(void *provctx, const char *uri,
+                          const char *propq, const OSSL_PARAM params[],
+                          OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg)
+{
+    struct file_ctx_st *ctx = NULL;
+
+    ctx = file_open(provctx, uri);
+    if (ctx == NULL)
+        return NULL;
+
+    if (propq) {
+        OPENSSL_free(ctx->_.file.propq);
+        ctx->_.file.propq = strdup(propq);
+        if (ctx->_.file.propq == NULL) {
+            ERR_raise(ERR_LIB_PROV, ERR_R_PROV_LIB);
+            free_file_ctx(ctx);
+            return NULL;
+        }
+    }
+
+    if (!file_set_ctx_params(ctx, params)) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_PROV_LIB);
+        free_file_ctx(ctx);
+        return NULL;
+    }
+
+    if (pw_cb) {
+        if (!file_setup_decoders(ctx)) {
+            ERR_raise(ERR_LIB_PROV, ERR_R_PROV_LIB);
+            free_file_ctx(ctx);
+            return NULL;
+        }
+
```

Comment:
I won't deny the possibility that the PKCS#11 provider needs to do more preparation early.  But, it also has objects stored in a more structured manner, so I fail to see why it would need `propq` specifically.

Reply:
Do note that the sanitizer CI builds all protest on this call...  something's not being cleaned up right

Reply:
The reason to pass propq is things like being able to deal with FIPs mode.
I would like to pass the fips=yes property given PKCS#11 3.x will support passing back FIPS indicators, which we may use to select if keys are going to be used or not based on their fips status.
But in order to do that we need to know if the app is passing any prop query to mcuk with the fips property (either to require or exempt it).

Reply:
That sounds...  a bit clumsy.  Or at least hacky.  After all, this is the same propq that's used to fetch the storemgmt implementation.

So it would as well make sense for the pkcs11 provider to provide two implementations, one with the property `fips_yes` and the other without.

All that takes is one `OSSL_DISPATCH` table for each implementation, that are almost exactly the same, except for two distinct `open_ex` functions, one that enables a FIPS flag in the loader ctx, and the other that disables it, and they call a common function that parses the URI and everything else.
That way, there's no need for you to parse the properties at all, as it's essentially already done for you.

Reply:
@levitte I am not sure which one is the right way to go yet, this property business is a bit too static at times assuming that a property is always true in absolute on the openssl side make things awkward when that is not true on the provider side.
Perhaps we need to think of static and dynamic properties at some point.
But coming back to this code I would rather risk not using a property passed in than not having it available and needing it.

Reply:
Note that I see issues with dynamically resolving interfaces in other places.

For example, to avoid issues in applications we try to do late binding of the pkcs11 module. Basically we want to activate it *only* if the applications starts using pkcs#11 URIs ... however we also implement operations that do not require keys like digests.

If I make digests available openssl ends up calling into the provider early and we end up having to load the module etc... and digests are called very early just to see if they are around or something.

However, if we do not provide the digest operation openssl will never call the provider for those operations anymore later, as it caches the fact they are available via default provider.

I tried to play with the cache option on load_operations but I got odd results and deferred till later.

Note that one goal here is to be able to make pkcs11-provider transparent to applications, so that a well written application that uses the modern store API doesn't even need to know that they are using the default or the fips provider via a pem file vs a pkcs#11 module and hw token, and things just work. This is kind of a goal for a distribution where you want to enable users to use HSMs or Smart Cards or other tokens like Yubikeys, with minimal user intervention and no application recompilation, across the whole system.

Reply:
The whole dynamic thing and all that sounds a bit off-topic in this PR, and should rather be raised as a separate issue.

The story about FIPS indicators for keys in PKCS#11 3.x sounds intriguing, though, and I suspect that may potentially come to affect more than just storemgmt.  I'm just getting more convinced that *with current provider interface design*, the most sane would still be two entries in the storemgmt `OSSL_ALGORITHM` array, something like this (you'll have to adapt names to your liking, of course):

``` C
const OSSL_ALGORITHM pkcs11_storemgmt[] = {
    { "pkcs11", "provider=pkcs11,fips=yes", pkcs11_storemgmt_fips_fns, "some FIPS desc..." },
    { "pkcs11", "provider=pkcs11,fips=no", pkcs11_storemgmt_fns, "some desc..." }
};

/* ... */

const OSSL_DISPATCH pkcs11_storemgmt_fips_fns = {
    {OSSL_FUNC_STORE_OPEN,                (void (*)(void)) pkcs11_store_open_fips},
    /*
     * All other functions are the same in pkcs11_storemgmt_fips_fns
     * and in pkcs11_storemgmt_fns, not shown here
     */
    {0, NULL}
};

const OSSL_DISPATCH pkcs11_storemgmt_fips_fns = {
    {OSSL_FUNC_STORE_OPEN,                (void (*)(void)) pkcs11_store_open},
    /*
     * All other functions are the same in pkcs11_storemgmt_fips_fns
     * and in pkcs11_storemgmt_fns, not shown here
     */
    {0, NULL}
};
```

`pkcs11_store_open_fips()` would then set a FIPS flag in the internal context structure, while `pkcs11_store_open()` wouldn't, and that flag can be used to try and match the FIPS indicator you mentioned in whatever way you find suitable.

Essentially, *this is what OpenSSL providers actually do* (except that we've placed fips=yes and fips=no implementations in separate providers, but we can ignore that in this discussion)

Reply:
Yes, this specific "issue" with the fips property can be solved that way.
But as you say let's put this aside and go back to the issue at hand.

I need a pin prompt at store open :-)

If you tell me unequivocally to drop propq from this open_ex call function signature I will.

Reply:
I've answer by the core_dispatch signature for it, hopefully for better clarity: https://github.com/openssl/openssl/pull/20131/files#r1231991315